accepted/00063-hub-iss.md

```diff
@@ -0,0 +1,250 @@
+
+The SQL script would be produced by a program that looks at the ISS Master's schema and, given a small set of starting tables (eg. `rhnchannel`) and starting rows given by a criterion (eg. `channel_id = 103` or `modified_date > '2019-01-01'`), figures out:
+
+- the graph of dependent/dependency tables (from the starting set)
+- for each table:
+  - all columns
+  - the primary key (if any), and columns involved
+  - the sequences feeding the primary key if numeric (heuristics with exceptions, or database schema naming rules, might be needed)
+  - the foreign keys (if any), and the columns involved
+  - the rows directly or indirectly connected to the starting row id
+
+With the above information, it should be possible to write the SQL script in a completely or almost-completely automated way, in particular:
+  - determining the right order of `INSERT` and `DELETE` statements given table dependencies
+  - determining which column values have to be `SELECT`ed, because they are foreign keys, and which are constant
+  - determining the conflict column set for `ON CONFLICT` clauses
+  - the actual row data to insert
+
+Expected benefits:
+  - the ISS programs need (potentially) no updating when the database structure changes
+  - supporting a new managed content type is as easy as specifying a new set of starting tables
+
+### Transport
+
+Both file and database data need to be transferred somehow from the ISS Master to the ISS Slave.
+
+Current ISS implementation offers:
+  - storage-based transfer (a directory with well-defined structure is created to host files and data in XML form) and
+  - API-based transfer: an XMLRPC request is made from ISS Slave to Server, and the same data set is sent over XMLRPC (http), streamed
+
+The new ISS implementation will eventually support both a storage-based transfer and an API-based transfer (not necessarily XMLRPC based). Details on the API-based transfer are not covered by this RFC, see "Limitations" below.
+
+### Triggering
+
+Current ISS implementation is ISS-Slave-triggered - the `mgr-inter-sync` command will trigger syncing (by default via the API-based transport).
+
+The new ISS implementation will be ISS-Master triggered, possibly via Salt states. The exact mechanism is left as an implementation detail.
+
+### Limitations
+
+The first implementation of this RFC will:
+  - be limited to software channels
+  - be limited to storage-based transfers
+
+## Implementation details
+
+- two commandline tools will be added, one to be used on an ISS Master and one on an ISS Slave
+  - output of the ISS Master tool will be a directory with files and a SQL script
+  - the ISS Slave tool will mainly copy files to correct locations and execute the SQL script
+  - triggering of the ISS Slave tool might happen via Salt, this is left as an implementation detail
+  - transfer proper could be done via `rsync`, which provides transparent compression and delta encoding
```

Comment:
If an external tool should be used maybe [`rclone`](https://rclone.org/) could be an alternative here: from my experience it has a much more user-friendly CLI than `rsync` (there seems to be a [python wrapper](https://pypi.org/project/python-rclone/) too), additional features (e.g. sync to and from network), support for all kinds of storage types, so API based transport could maybe even be implemented as an Uyuni plugin.